### ISSUE TITLE
Clean is slower in sbt 1.3.0

### ISSUE THREAD

**Issue by lbialy** [sbt/sbt#4972]
## steps
sbt version: 1.3.0-RCx

## problem
sbt is getting gradually slower instead of getting faster. I am currently optimizing build in my work project (~60 subprojects) and I was doing some trivial comparisons of performance in 1.3 as a counterargument to migration to gradle. I have written a small script to roughly measure differences with bash's `time`:

```bash
#!/bin/bash

BRANCHES=("sbt-1.2.8" "sbt-1.3.0-RC3" "sbt-1.3.0-RC4" "sbt-1.3.0-RC4-turbo")

cd WORK_PROJECT

for BRANCH in ${BRANCHES[@]}
do
        git checkout "$BRANCH"
        exec 3>&1 4>&2
        result=$( { time sbt clean "some-subproject/test" 1>/dev/null 2>/dev/null; } 2>&1 )
        exec 3>&- 4>&-
        echo "Results for $BRANCH:"
        echo $result
        echo ""
done
```

There are 4 branches as you can see: 
* 1.2.8 with sbt-coursier plugin enabled
* 1.3.0-RC3
* 1.3.0-RC4
* 1.3.0-RC4 with turbo mode enabled

Turbo mode can't be enabled for RC3 due to a bug that crashes some of the tests.

Results are as follow:

```
λ ./sbt_perftest.sh                                                                                                                                                                                 
Already on 'sbt-1.2.8'
Results for sbt-1.2.8:
real 0m32.584s user 2m3.421s sys 0m2.241s

Switched to branch 'sbt-1.3.0-RC3'
Results for sbt-1.3.0-RC3:
real 0m44.143s user 2m47.532s sys 0m3.471s

Switched to branch 'sbt-1.3.0-RC4'
Results for sbt-1.3.0-RC4:
real 0m57.424s user 3m18.849s sys 0m4.514s

Switched to branch 'sbt-1.3.0-RC4-turbo'
Results for sbt-1.3.0-RC4-turbo:
real 0m37.382s user 2m7.711s sys 0m3.870s
```

## expectation
sbt not getting slower?

## notes
**The biggest offender seems to be the `clean` task which takes:**
* 0 seconds on 1.2.8
* 4 seconds on 1.3.0-RC3
* 9 seconds on 1.3.0-RC4
* 9 seconds on 1.3.0-RC4 with turbo mode 

test target also differs (it's an empty scalatest target, there are no test suites in this subproject):
* 12 seconds on 1.2.8
* 12 seconds on 1.3.0-RC3
* 14 seconds on 1.3.0-RC4
* 13 seconds on 1.3.0-RC4 with turbo mode

Those differences are small but I'm not sure they can be discarded as runtime artifacts (also, it's kind of a big deal if build tool has variable overhead of +/- few seconds).

**Comment by eatkins** [comment]
Yes. You are right that in many cases clean is slower than it was before. Thank you for making an issue. I'd been meaning to make one.

**Comment by lbialy** [comment]
I strongly disagree with the change of name and scope, let me elaborate:

```bash
λ cat ./sbt_perftest.sh                                                                                                                                                                            
#!/bin/bash

BRANCHES=("sbt-1.2.8" "sbt-1.3.0-RC3" "sbt-1.3.0-RC4" "sbt-1.3.0-RC4-turbo")

cd WORK_PROJECT

for BRANCH in ${BRANCHES[@]}
do
        git checkout "$BRANCH"
        sbt clean > /dev/null 2>&1
        exec 3>&1 4>&2
        result1=$( { time sbt "some-subproject/test" 1>/dev/null 2>/dev/null; } 2>&1 )
        exec 3>&- 4>&-
        sbt clean > /dev/null 2>&1
        exec 3>&1 4>&2
        result2=$( { time sbt "some-subproject/test" 1>/dev/null 2>/dev/null; } 2>&1 )
        exec 3>&- 4>&-
        sbt clean > /dev/null 2>&1
        exec 3>&1 4>&2
        result3=$( { time sbt "some-subproject/test" 1>/dev/null 2>/dev/null; } 2>&1 )
        exec 3>&- 4>&-
        echo "Results for $BRANCH:"
        echo $result1
        echo $result2
        echo $result3
        echo ""
done
```

As you can see clean step is now separated and not measured. I am also measuring 3 runs per version. Here's the result:

```
./sbt_perftest.sh                                                                                                                                                                                
Switched to branch 'sbt-1.2.8'
Results for sbt-1.2.8:
real 0m23.329s user 1m22.461s sys 0m1.784s
real 0m24.254s user 1m27.477s sys 0m1.700s
real 0m25.032s user 1m30.185s sys 0m1.890s

Switched to branch 'sbt-1.3.0-RC3'
Results for sbt-1.3.0-RC3:
real 0m28.118s user 1m42.418s sys 0m2.082s
real 0m27.774s user 1m39.593s sys 0m2.155s
real 0m28.485s user 1m45.187s sys 0m1.993s

Switched to branch 'sbt-1.3.0-RC4'
Results for sbt-1.3.0-RC4:
real 0m30.059s user 1m49.518s sys 0m2.197s
real 0m30.011s user 1m48.425s sys 0m2.200s
real 0m30.682s user 1m52.706s sys 0m2.139s

Switched to branch 'sbt-1.3.0-RC4-turbo'
Results for sbt-1.3.0-RC4-turbo:
real 0m28.992s user 1m45.262s sys 0m1.930s
real 0m26.742s user 1m38.779s sys 0m1.885s
real 0m27.101s user 1m39.280s sys 0m1.971s
```

**Comment by eatkins** [comment]
https://github.com/sbt/sbt/issues/4974

**Comment by eatkins** [comment]
I understand that you are frustrated by the 3-5 second startup penalty in 1.3.0x, but this is unlikely to be a high priority issue. sbt is not optimized for cold startup time. The clean issue will be fixed soon.

**Comment by lbialy** [comment]
I am aware that sbt is not optimized for cold startup time, but sadly, there is no _sane_ way to make CI pipelines with sbt running as daemon/server. I've already tried using sbt with https://github.com/cb372/sbt-client but sadly RPC is not mature (for instance from what I recall errors / exceptions are not published to RPC output channel and only show up in console in which sbt server is running, moreover: server is not stable and after few hours of usage crashes with metaspace OOMs even if class unloading is enabled).

**Comment by lbialy** [comment]
On the other hand, not to sound entirely negatively, here are the results of our whole test suite running with different versions of sbt:

```
./sbt_perftest.sh                                                                                                                                                                                
Switched to branch 'sbt-1.2.8'
Results for sbt-1.2.8:
real 12m15.694s user 33m28.293s sys 0m59.584s
real 12m12.545s user 34m6.221s sys 1m0.628s
real 12m21.325s user 33m35.848s sys 1m0.142s

Switched to branch 'sbt-1.3.0-RC3'
Results for sbt-1.3.0-RC3:
real 10m41.446s user 29m33.266s sys 0m58.289s
real 10m45.206s user 30m30.798s sys 0m57.808s
real 0m58.598s user 3m23.073s sys 0m4.577s

Switched to branch 'sbt-1.3.0-RC4'
Results for sbt-1.3.0-RC4:
real 10m40.387s user 30m3.883s sys 0m57.090s
real 10m31.102s user 29m10.926s sys 0m56.823s
real 40m27.102s user 30m9.295s sys 1m14.206s

Switched to branch 'sbt-1.3.0-RC4-turbo'
Results for sbt-1.3.0-RC4-turbo:
real 10m16.618s user 29m18.842s sys 0m55.158s
real 10m26.260s user 29m7.081s sys 0m56.497s
real 40m33.261s user 31m35.889s sys 1m14.768s
```

There seems to be some degree of nice improvement with ~2 minutes shaved off in full runs but as you can see tests are not stable under RC3 and RC4. I will investigate this instability today.

**Comment by eed3si9n** [comment]
@eatkins 

> I understand that you are frustrated by the 3-5 second startup penalty in 1.3.0x, but this is unlikely to be a high priority issue. sbt is not optimized for cold startup time. The clean issue will be fixed soon.

It's true that sbt hasn't always been optimized for cold startup time, but we have tried to reduce the load time in different attempts like (https://github.com/sbt/sbt/pull/4067 etc), so any worsening in load time should be considered a regression. Since load time affects all users, I personally think it should be given higher priority than most feature enhancements.

I guess we can continue in https://github.com/sbt/sbt/issues/4974.

**Comment by eatkins** [comment]
@lbialy re startup: see the discussion in https://github.com/sbt/sbt/issues/4974.

With respect to your tests sometimes taking 40 minutes, are you by chance using scala 2.12.9? There was a compile performance regression: https://github.com/scala/bug/issues/11667.

**Comment by lbialy** [comment]
Nope, 2.11.12 all the way, it's a spark 2.3 based project.

That 40 minutes might be a flaky test running with embedded kafka, but I'm more worried about those short ones - they seem to be compiler classpath failures in RCx. I will be opening a new issue for those.

**Comment by eed3si9n** [comment]
I'm going to close this as fixed by #4978.
Feel free to open further issues if there are any other areas of perf regression.
